Tests/TestingApps/PaywallsTester/PaywallsTester/Data/OfferingsPaywallsViewModel.swift

```diff
@@ -84,32 +110,41 @@ extension OfferingsPaywallsViewModel {
         var paywalls: [PaywallsResponse.Paywall]
 
         func paywallsByOffering() -> [OfferingPaywall] {
-            let paywallsByOfferingID = Set(self.paywalls).dictionaryWithKeys { $0.offeringID }
+            let paywallsByOfferingID = self.paywalls.dictionaryWithKeys { $0.offeringID }
 
             var offeringPaywall = [OfferingPaywall]()
             for offering in self.offerings {
                 if let paywall = paywallsByOfferingID[offering.id] {
-                    offeringPaywall.append(OfferingPaywall(offering: offering, paywall: paywall))
+                    let rcOffering = paywall.convertToRevenueCatPaywall(with: offering)
+                    offeringPaywall.append(OfferingPaywall(offering: offering, paywall: paywall, rcOffering: rcOffering))
                 }
             }
 
             return offeringPaywall
         }
 
+        func offeringsWithoutPaywalls() -> [OfferingsResponse.Offering] {
+            let paywallsByOfferingID = self.paywalls.dictionaryWithKeys { $0.offeringID }
+
+            return self.offerings.filter { paywallsByOfferingID[$0.id] == nil }
+        }
+
     }
 
     @MainActor
-    private func showPaywallForID(_ id: String) {
-        switch self.offeringsPaywalls {
-        case let .success(data):
-            if let dataForRequestedID = data.first(where: { $0.offering.id == id }) {
-                let newRCOffering = dataForRequestedID.paywall.convertToRevenueCatPaywall(with: dataForRequestedID.offering)
+    private func showPaywallForID(_ id: String, mode: PaywallViewMode = .default) {
+        switch self.state {
+        case .notloaded:
+            Self.logger.log(level: .info, "Could not show paywall for id \(id), data not loaded.")
+            self.presentedPaywall = nil
```

Comment:
from a user's perspective, are we displaying any kind of an error message? A log wouldn't be enough if the user isn't running it from Xcode

Reply:
I haven't put in any user-facing error, good idea, will do so as part of another PR. I made [a linear ticket](https://linear.app/revenuecat/issue/PWL-458/user-facing-error-dialogs).